### DISCUSSION TITLE
Include dellemc.powerflex collection

### DISCUSSION THREAD

**Discussion by khareRajshree** [ansible-collections/ansible-inclusion#33] [Resolved reviews]
Intent is to include dellemc.powerflex collections in Ansible 5.0.0 community release. Please validate and provide your feedback for the same.

Collection Link: https://galaxy.ansible.com/dellemc/powerflex
Repo: https://github.com/dell/ansible-powerflex

**Comment by tadeboro** [comment]
@khareRajshree  Please read the [README](https://github.com/ansible-collections/ansible-inclusion/blob/main/README.md) file and supply all required information. And please do go over [the checklist](https://github.com/ansible-collections/overview/blob/main/collection_checklist.md) before asking for inclusion review. Thank you.

**Reply by khareRajshree** [reply]
Thanks @tadeboro for providing this information, we will go through the check list and will get back to you.

**Comment by tadeboro** [comment]
Thank you for your submission, @khareRajshree.

The items in the checklist can be in one of the following states:

  * Items with a checkmark and no attached comment are OK and need no intervention from your side.
  * Items with a checkmark that contain a comment will not block the inclusion process, but by following the advice given in the comment, the collection's quality can be improved.
  * Items with no checkmark and no extra details are waiting for evaluation.
  * Items with no checkmark that contain a comment need your attention.

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [ ] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
      **MUST FIX:** This information is not available in the collection. Some examples: https://sensu.github.io/sensu-go-ansible/release_policy.html, https://github.com/ansible-collections/community.sops#releasing-versioning-and-deprecation
- [ ] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
      **MUST FIX:** Code of conduct is missing from the collection.
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [ ] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
      **MUST FIX:** https://github.com/dell/ansible-powerflex/blob/main/plugins/doc_fragments/dellemc_powerflex.py is missing a license header.
      **MUST FIX:** `galaxy.yml` contains both `license` and `license_file` options. One of them must go.
- [ ] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
      **SHOULD FIX:** https://github.com/dell/ansible-powerflex/blob/e494bc2015eab041700a63fb789c534addb46881/plugins/doc_fragments/dellemc_powerflex.py#L47 is not needed since boolean arguments only take true or false values.
      **SHOULD FIX:** [Metadata](https://github.com/dell/ansible-powerflex/blob/e494bc2015eab041700a63fb789c534addb46881/plugins/modules/dellemc_powerflex_device.py#L10-L12) is ignored by ansible-base and ansible-core and can be removed.
      **MUST FIX:** Modules do not document the `PyPowerFlex` python requirement. The readme does contain this information, but reference docs do not.
- [ ] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [ ] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
      **MUST FIX:** https://github.com/dell/ansible-powerflex/blob/main/plugins/modules/dellemc_powerflex_gatherfacts.py is an info module, not a facts module (it does not actually set facts).
  - [ ] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
      **MUST FIX:** As far as I can tell, all modules are a mix of info and regular modules (when the `state` argument is set to `present`, a certain combination of parameters makes modules behave as info modules). This functionality should not be part of the normal module.
  - [ ] `check_mode` is supported in all `_info` and `_facts` modules
      **MUST_FIX:** https://github.com/dell/ansible-powerflex/blob/main/plugins/modules/dellemc_powerflex_gatherfacts.py does not support check mode.
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [ ] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
      **MUST FIX:** Since Ansible 5 will depend on `ansible-core-2.12`, the collection also needs to be tested against stable-2.12 branch.
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [ ] all CI tests run regularly (nightly, or at least once per week)
      **MUST FIX:** There is no scheduled CI run.

**Other:**

- **MUST FIX:** `galaxy.yml` file points to tags in the `repository`, `documentation`, and `homepage` sections.

**Reply by tadeboro** [reply]
I finished my initial review. The collection is not ready for inclusion in the current state.

**Comment by khareRajshree** [comment]
Hi  @tadeboro , we have updated the collections based on the checklist, I request you to please review it once again.

Ansible Galaxy: https://galaxy.ansible.com/dellemc/powerflex
GitHub: https://github.com/dell/ansible-powerflex
Issues Tracker: https://github.com/dell/ansible-powerflex/issues

Thanks,
Rajshree

**Reply by khareRajshree** [reply]
Hi @tadeboro, a gentle reminder.

**Comment by anupamaloke** [comment]
@tadeboro, @felixfontein, appreciate your help and review. Now that ansible community package 5.0.0 is out, we are submitting the ansible-powerflex collection for inclusion in ansible community package 6.0.0. Do we need to create a  new request or we can continue this discussion thread itself?

**Reply by gundalow** [reply]
> @tadeboro, @felixfontein, appreciate your help and review. Now that ansible community package 5.0.0 is out, we are submitting the ansible-powerflex collection for inclusion in ansible community package 6.0.0. Do we need to create a new request or we can continue this discussion thread itself?

Continue to use this request.
Inclusion reviews are generally done by members of the Ansible Community in their spare time.

**Reply by tadeboro** [reply]
Currently, we are still targeting Ansible 6.0.0 with inclusions. But things might change once the voting on https://github.com/ansible-community/community-topics/issues/63 is over. If the proposed idea passes the vote, we will try to get this collection included in the Ansible community package as soon as possible and will not wait for 6.0.0.

As for the silence with reviews: the sad truth is that there is not much happening right now. There are multiple reasons for this (people being busy with other things, etc.), but one of the reasons is also the lack of urgency due to Ansible 6 being almost half a year away. And the linked proposal tries to change this and help speed things up a bit.

**Reply by anupamaloke** [reply]
@gundalow, @tadeboro, thank you so much for your prompt reponse. I like the new proposal that allows inclusion of the collection in the minor releases as well and voted for it :-)

**Reply by anupamaloke** [reply]
@felixfontein, @tadeboro, @gundalow, now that voting is completed and ansible inclusion process is simplified, I just wanted to check when we can expect the collection to be included in the ansible community release. I know you guys have been very busy, so there is no rush.

**Comment by anupamaloke** [comment]
@tadeboro, just checking to make sure that this is also there in your list of collections that need to be included in the next minor release :-)

**Reply by tadeboro** [reply]
I am not part of the Ansible community anymore, so you will need to find someone else to push this forward.

**Reply by anupamaloke** [reply]
@tadeboro, thanks a lot for your update.

**Comment by Andersson007** [comment]
@khareRajshree @anupamaloke have all the issues from tadeboro's review been fixed?

If yes, I'll change the category to the `Second review needed` and will announce via Bullhorn that the collection needs another review.
As soon as there's a volunteer from the Community or @ansible-collections/steering-committee , after the collection is reviewed the second time, it'll be included after that.

**Reply by anupamaloke** [reply]
@Andersson007, thank you for your response. Yes, we have fixed all the issues from tadebro's review.

**Reply by Andersson007** [reply]
@anupamaloke i don't think all the issues have been fixed.. you could start fixing them while I'm reviewing the collection, no need to wait

**Comment by Andersson007** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

Note for reviewers: If you don't know how to check any of the points below, please ask maintainers of the collection you're reviewing or a [Steering Committee member](https://docs.ansible.com/ansible/devel/community/steering/community_steering_committee.html#current-steering-committee-members) for clarifications in comments of corresponding inclusion discussion.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
  **MUST FIX**: (copied from tadaboro's review) This information is not available in the collection. Some examples: https://sensu.github.io/sensu-go-ansible/release_policy.html, https://github.com/ansible-collections/community.sops#releasing-versioning-and-deprecation
  **MUST FIX**: Also add info that you follow SemVer.
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
  **MUST FIX**: (copied from tadaboro's review) Code of conduct is missing from the collection.
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
  **MUST FIX**: (copied from tadaboro's review) doc_fragments/dellemc_powerflex.py is missing a license header.
  **MUST FIX**: (copied from tadaboro's review) galaxy.yml contains both license and license_file options. One of them must go.
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  **MUST FIX**:  (copied from tadaboro's review) Modules do not document the PyPowerFlex python requirement. The readme does contain this information, but reference docs do not.
  **MUST FIX**: Use proper markup (see the doc guidelines "Module format and documentation"

for example, in the following fragment, `sds_id` must be `I(sds_id)`. I can see many of other doc formatting issues across the files including in `doc_fragments/powerflex.py`. Please recheck all the files and make the doc satisfying the doc standards:
```
  38   sds_name:
  39     description:
  40     - The name of the SDS.
  41     - Required while adding a device.
  42     - Mutually exclusive with sds_id.
```

- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
  **MUST FIX**: As collection does not support all the [required Python versions](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#other-environment), please add to `doc_fragments/powerflex.py requirements:` the Python versions the collection supports.
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core
- [x] public plugins, roles and playbooks do not use files outside of `meta/`, `plugins/`, `roles/`, and `playbooks/`

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files created by development tools. Temporary files MUST be added to `.gitignore`.

**Tests:**

Note for reviewers: If you don't know how to check the points below, please ask maintainers of the collection you're reviewing how you can do it.
* In most cases, CI is set up via GitHub Actions.
* Check `.yml` files in the `.github/workflows` directory. There must be at least the `sanity` section under `jobs` containing the `ansible-test sanity` command running against all supported ansible-core versions that must be also listed there, for example, `- stable-2.11, - stable-2.12, - stable-2.13`.
* Check workflow runs by clicking the `Actions` tab in the repository's page - you're interested in `Scheduled` runs, runs against release commits and runs against pull requests.
* If there are no workflows in the `Actions` tab, ask the collection maintainers how CI is implemented.

- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it MUST not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests MUST run against every pull request
- [x] all CI tests MUST run regularly (nightly, or at least once per week)
  **MUST FIX**: they do not run regularly, please add scheduled run directive to your GHA matrix
- [x] sanity tests MUST run against a commit that releases the collection; if they don't pass, the collection won't be released

**Reply by Andersson007** [reply]
@anupamaloke I've done with the first round of my review. Could you please take a look?
Also the issues from it can be relevant to your [second inclusion request](https://github.com/ansible-collections/ansible-inclusion/discussions/32). If you take them into consideration there, before someone else starts its review, it'll speed up the progress, I'm sure.

thanks

**Reply by anupamaloke** [reply]
@Andersson007, thank you for reviewing this. I will get started on getting these comments fixed.

**Comment by kuttattz** [comment]
Hi @felixfontein, @Andersson007 , We have incorporated the latest feedbacks suggested and updated the repo - https://github.com/dell/ansible-powerflex. Requesting you to review it and let us know if the code is compliant now. 
Thanks

**Reply by Andersson007** [reply]
@kuttattz thanks for pinging, I'll try to take a look as soon as I have time

**Reply by Andersson007** [reply]
@kuttattz hi, looks almost ready for inclusion, thanks for fixing the stuff!
Two last points from me:
- [x] I created a PR do make the docs adhering to the latest doc standards (a decision has been made recently about the boolean representation), needs to get merged https://github.com/dell/ansible-powerflex/pull/21
- [x] I don't see CI runs (sanity + units) against my PR, should be fixed

Please let me know when they are done

**Reply by Jennifer-John** [reply]
Hi @Andersson007 the CI runs on [Change doc formatting for inclusion in Ansible PR](https://github.com/dell/ansible-powerflex/pull/21) is approved to run and they were successful and the PR is merged. Thank you!

**Reply by Andersson007** [reply]
@Jennifer-John ah, sorry, i see now, thanks!

@felixfontein @mariolenz @gotmax23 Tadej's PR wasn't finished but as far as I can see the maintainers fixed things he spotted. Would you like to take a quick look and confirm this?

**Comment by mariolenz** [comment]
Looks mostly OK to me. However, I'm not sure if we should accept specific RHEL versions as prerequisites. I think Python and ansible-core versions are OK, but why a specific distro?

That said, not supporting 2.14 is a problem since Ansible 7 is based on this. It looks like the CI tests ~agaisnst~ against 2.14, so it might be just an outdated documentation in the [README](https://github.com/dell/ansible-powerflex/blob/10524b78e30e8a26d5c4b2323c180e3aa6145a92/README.md).

**Reply by felixfontein** [reply]
Hmm, asking for specific OS versions without a good reason sounds like a no-go to me. I would *guess* that these are not hard requirement but more statements on what has been tested, but if that's the case the wording should really be changed.

And yes, ansible-core 2.14 not being mentioned in the README is odd.

**Reply by Jennifer-John** [reply]
Thank you @mariolenz @felixfontein for the review. Yes the readme will be updated with support for 2.14 as part of the release on 13th dec as we were still working on the regression tests on 2.14. As you have rightly mentioned, these are the tested RHEL versions. We will try to re-word or remove it from the pre-requisites. Once we have the release, I will post here the updated links to the readme.
Thanks,
Jennifer.

**Reply by Andersson007** [reply]
@Jennifer-John thanks for the fast feedback, please let us know when we can take another look

**Reply by Jennifer-John** [reply]
sure @Andersson007, thank you!

**Comment by Jennifer-John** [comment]
Hi @Andersson007 , @felixfontein @mariolenz Thank you for the feedback. We have updated the readme to remove the specific RHEL versions and 2.14 is now included in supported versions. The latest changes are released and are part of the ansible galaxy release 1.5.0 https://galaxy.ansible.com/dellemc/powerflex. 
Could you please help re-review https://github.com/dell/ansible-powerflex for inclusion? 

Thanks,
Jennifer

**Comment by mariolenz** [comment]
The README looks OK to me now. Additionally, it looks like they test against [ansible-core 2.14](https://github.com/dell/ansible-powerflex/actions/runs/3691875015/jobs/6250262667).

@Andersson007 @felixfontein I think this collections is fit to be added to the community package.

**Reply by Andersson007** [reply]
@mariolenz feel free to submit a topic similar to https://github.com/ansible-community/community-topics/issues/163 (or I can after I'm back from PTO)

**Reply by mariolenz** [reply]
> @mariolenz feel free to submit a topic similar to [ansible-community/community-topics#163](https://github.com/ansible-community/community-topics/issues/163) (or I can after I'm back from PTO)

FYI @Andersson007: ansible-community/community-topics#177

**Comment by Jennifer-John** [comment]
Thank you @mariolenz for the review, @Andersson007 @felixfontein i also want to bring to your notice post the last review apart from the readme changes, there have also been code changes (which is part of the released collection) for which we have incorporated the community guidelines.
Thanks,
Jennifer.

**Comment by ssbarnea** [comment]
-1 CG should not grow, especially with code that is not "general"-enough.

**Reply by Jennifer-John** [reply]
@ssbarnea could you share some more details on this comment for us to check and implement any required improvements on the code? Thank you!

**Reply by gotmax23** [reply]
@ssbarnea said:

> -1 CG should not grow, especially with code that is not "general"-enough.

This ticket has nothing to do with community.general. It's is a request to add a new collection to the `ansible` package.

**Reply by Andersson007** [reply]
some time ago this topic was raised and we got feedback from the community in https://github.com/ansible-community/community-topics/issues/82 and on reddit: the majority of participants from the community is happy to see any collections (even not general) satisfying the collection requirements being included in the package.

**Reply by Jennifer-John** [reply]
Hi @ssbarnea @gotmax23 @Andersson007 are we good with the community inclusion for the collection or do you see any further improvements required?

Thanks,
Jennifer.

**Reply by Andersson007** [reply]
@Jennifer-John according to https://github.com/ansible-community/community-topics/issues/177, it's scheduled to happen on 2023-01-12

**Reply by Andersson007** [reply]
@ssbarnea if you think we should formally (update: in a formal way i mean) vote on the inclusion, please put a not in the topic ^

**Reply by Jennifer-John** [reply]
Thank you @Andersson007 for the update

**Reply by mariolenz** [reply]
@Jennifer-John I've opened ansible-community/ansible-build-data#193 but it's not merged yet. But I hope someone will do soon.

cc @Andersson007

**Reply by Andersson007** [reply]
Done, thanks everyone!:)

**Reply by Jennifer-John** [reply]
Thank you @mariolenz @Andersson007! Thank you team for the reviews!

Thanks,
Jennifer
